cmd-run: switch to qemu image layering and injection of config.ign in to the /boot

jcajka opened the pull request:
This will make ```cosa run``` work on platforms that don't support qemu_fw_cfg, although it is kind of useless as there is no ignition config. But IMHO this clearly document it and make us able to smoke test if it boots at all locally.

jlebon commented on src/cmd-run:
```diff
@@ -173,7 +173,12 @@ EOF
+case $arch in
+x86_64|aarch64) set -- -fw_cfg name=opt/com.coreos/config,file="${IGNITION_CONFIG_FILE}" "$@" ;;
+# no qemu_fw_cfg on these architectures just run the image without the configuration
+ppc64le|s390x) set -- "$@" ;;
```
I like the explicit `case` to make it clear it's not supported on those platforms. Though we can just do `;;` here, right?

r4f4 replied:
Maybe some message instead in case someone is trying to test on `ppc64le` and wondering why the ignition file is not working?

cgwalters commented:
We can change `cmd-run` to create a temporary qcow2 overlay, and use libguestfs to inject ignition into `/boot` instead.

As a bonus we also need this for making it easy to run a `metal-uefi` image e.g.

jcajka commented:
@cgwalters I like that idea. I would be happy to look in that direction. Do you have any tips for start? I guess qemu-img and guestfish are tool to go to, right?

jcajka commented:
@cgwalters could you also elaborate on the injection in to the /boot? Does the location of the ignition file needs to be specified as kernel parameter or is there implicit location that is searched?

dustymabe commented:
> @cgwalters I like that idea. I would be happy to look in that direction. Do you have any tips for start? I guess qemu-img and guestfish are tool to go to, right?

Yeah those are the right tools. Let's be careful to create a new disk image or a snapshot before we inject the file so as to not modify the actual built image which you might share with someone else later and you want it to be pristine.  


> @cgwalters could you also elaborate on the injection in to the /boot? Does the location of the ignition file needs to be specified as kernel parameter or is there implicit location that is searched?

Nope. If the file exists in `/boot/ignition/config.ign` and it is a firstboot (i.e. ignition gets run) then it will get picked up. https://github.com/coreos/ignition-dracut/blob/d4e8899c91302a887e157fe2142bc1342f473c89/dracut/30ignition/ignition-setup.sh

jcajka commented:
Just pushed first version for comments. There is a known issue on ppc64le as it is using different default console, for that reason I have added default password for the core user, "coreos". I have tested it both on x86_64(fw_cfg and injection works) and ppc64le(injection works). I guess  with this change https://github.com/coreos/ignition/issues/666 is kind of non issue as it is used mainly by ```cosa run```, right?

jcajka commented:
@dustymabe I have fixed the auto-login issue and dropped the password. I believe it is ready for review/merge. 

@r4f4 would you mind giving a spin on aarch64?

r4f4 commented on src/cmd-run:
```diff
@@ -86,6 +105,12 @@ if [ -z "${VM_DISK}" ]; then
 fi
 
+if [ "${VM_LAYERED}" = 1 ]; then
+    LAYRED_DISK=tmp/layred-qemu.qcow2
```
Not really important but a typo in "layered"

r4f4 commented:
> @r4f4 would you mind giving a spin on aarch64?

build on the way.

r4f4 commented:
I was able to generate an image and run it as before with this patch on `aarch64`.

cgwalters commented on src/cmd-run:
```diff
@@ -173,7 +203,26 @@ EOF
-set -- -fw_cfg name=opt/com.coreos/config,file="${IGNITION_CONFIG_FILE}" "$@"
+
+if [ "${VM_INJECT}" = 1 ]; then
+    export LIBGUESTFS_BACKEND=direct
```
We also have a `libguestfish.sh`.

cgwalters commented on src/cmd-run:
```diff
@@ -173,7 +203,26 @@ EOF
+    rm -rf tmp/ignition
+fi
+
+case $arch in
```
Shouldn't this be `if [ "${VM_INJECT}" = 0 ]`?

cgwalters commented on src/cmd-run:
```diff
@@ -173,7 +203,26 @@ EOF
+    BOOT=$(guestfish --remote -- findfs-label boot)
+    guestfish --remote -- mount "${BOOT}" /
+    mkdir -p tmp/ignition
+    cat ${IGNITION_CONFIG_FILE} > tmp/ignition/config.ign
```
Why  `cat` instead of e.g. `cp --reflink=auto "${IGNITION_CONFIG_FILE} tmp/ignition/config.ign`?

cgwalters replied:
Another concern I have here is around SELinux labeling; any files we create via libguestfs need to be labeled correctly.  This *might* work today accidentally?  Regardless we'll likely need to beef up our support for that in our internal libguestfs helper.

jcajka replied:
@cgwalters what should be the correct selinux label for the ignition file?

cgwalters replied:
That is...a tricky topic.  I'll take a look at adding a primitive to our internal `libguestfs.sh` for this - see related code in https://github.com/cgwalters/qcow2-to-vagrant/blob/master/qcow2-to-vagrant

cgwalters replied:
OK I tossed up https://github.com/coreos/coreos-assembler/pull/444 but I am having other apparent troubles with my local build after the Ignition v3 switch, haven't debugged that yet.

cgwalters replied:
And here's my work so far on top of this PR:

```
From 8a48c8ca33c2ac34f43e8a15b3fcd148f052d028 Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Thu, 28 Mar 2019 00:24:56 +0000
Subject: [PATCH] wip

---
 src/cmd-run | 17 ++++++-----------
 1 file changed, 6 insertions(+), 11 deletions(-)

diff --git a/src/cmd-run b/src/cmd-run
index d629b41c..3dc59212 100755
--- a/src/cmd-run
+++ b/src/cmd-run
@@ -10,6 +10,8 @@ set -euo pipefail
 dn=$(dirname "$0")
 # shellcheck source=src/cmdlib.sh
 . "${dn}"/cmdlib.sh
+# shellcheck source=src/libguestfish.sh
+. "${dn}"/libguestfish.sh
 
 VM_DISK=
 VM_MEMORY=2048
@@ -190,17 +192,10 @@ EOF
 fi
 
 if [ "${VM_INJECT}" = 1 ]; then
-    export LIBGUESTFS_BACKEND=direct
-    eval "$(guestfish --listen)"
-    guestfish --remote -- add ${VM_DISK}
-    guestfish --remote -- run
-    BOOT=$(guestfish --remote -- findfs-label boot)
-    guestfish --remote -- mount "${BOOT}" /
-    mkdir -p tmp/ignition
-    cat ${IGNITION_CONFIG_FILE} > tmp/ignition/config.ign
-    guestfish --remote -- copy-in "$(realpath tmp/ignition)" /
-    guestfish --remote -- exit
-    rm -rf tmp/ignition
+    coreos_gf_run_mount "${VM_DISK}"
+    coreos_gf upload ${IGNITION_CONFIG_FILE} /boot/config.ign
+    coreos_gf_relabel /boot/config.ign
+    coreos_gf_shutdown
 fi
 
 case $arch in
-- 
2.20.1

```

jcajka replied:
Yup, mine biggest issue atm is the after forward porting my changes seem to not work anymore..., might be the selinux but yeah I need to dig deeper again...

jcajka replied:
Forwardported and some fix ups and it works again for me

cgwalters commented:
Just some relatively minor things, overall LGTM and thanks a lot for doing this!

dustymabe commented:
Thanks @jcajka - will run through and test this once I'm done with the vmware PR

dustymabe commented:
ok i've finally had a chance to look at this (sorry was concentrating on the vmware work). Overally here are my comments:

- I think I'd prefer to just use the same approach for all architectures. This means we would default to creating a layered disk image and injecting the ignition config.
- It doesn't look like we are cleaning up the layered disk image after the VM is done booting. We should do that right?
- We have an option already (`VM_PERSIST`) and a use of the `-snapshot` arg for  `qemu`. Let's just get rid of that in favor of the layered image approach. In the case someone does want it to persist (is anyone using this??) then we just wouldn't clean up the layered image I guess or we'd merge the layered image back into the base image.
- There are quite a few places (including in variable names) where `layered` is mispelled `layred`. Let's make sure we don't introduce any bugs by having those typos.

^^ all of these points are up for debate, just my `$.02`

r4f4 commented on src/cmd-run:
```diff
@@ -112,6 +137,11 @@ kernel.printk = 3 4 1 7
 )
 
+case $arch in
+    x86_64|aarch64|s390x) VM_TERMINAL="ttyS0";;
```
Can you make `aarch64`'s terminal `"ttyAMA0"`?

jcajka replied:
Sure, I will add that. It is the default console in the qemu, right? I haven't got around to evaluate this there yet.

r4f4 replied:
Yes. Thank you.

jcajka commented:
@dustymabe Ok, I haven't go the courage to propose that from the start. I will create the next iteration as replacement for the fw_cfg across the board. I hope to have it by tomorrow.

Other issue is to remove the layered image VM image as we are loosing control with the exec. I'm not sure what would be the correct course of action here. AFAIK we could just fail if the preserve flag is set and overwrite otherwise or a like(other way around as default).

ausil commented:
@cgwalters can you please take another look at the current patch

jcajka commented:
@cgwalters @ausil I'm atm moment working on incorporating the cleanups proposed by @cgwalters for some reason it doesn't work for me. I'm currently investigating why.

jcajka commented:
To be more specific this is the code that doesn't work. I have verified that the .ign file is create in the boot partition, but it seems that it is not processed by the ignition.

bgilbert commented on src/cmd-run:
```diff
@@ -158,7 +170,11 @@ EOF
-set -- -fw_cfg name=opt/com.coreos/config,file="${IGNITION_CONFIG_FILE}" "$@"
+
+coreos_gf_run_mount "${VM_DISK}"
+coreos_gf upload ${IGNITION_CONFIG_FILE} /boot/config.ign
```
This should be `/boot/ignition/config.ign`.

jcajka replied:
Doh...

jcajka commented:
I hope this will work once the ignition is fixed, version of it is synced with the assembler... :/

r4f4 commented:
I tried this PR out on `x86_64` machine after rebasing to latest cosa and got the following error message:
```
+ coreos_gf upload /proc/self/fd/3 /boot/ignition/config.ign
+ guestfish --remote -- upload /proc/self/fd/3 /boot/ignition/config.ign                                                           
libguestfs: error: upload: /boot/ignition/config.ign: No such file or directory                                                    
+ _coreos_gf_cleanup
+ guestfish --remote -- exit
+ :
```

jlebon commented:
Probably missing a `coreos_gf mkdir-p /boot/ignition` ?

r4f4 commented:
Yes, adding it just before the `upload` command worked.

jcajka commented:
@jlebon @r4f4 checking that out atm, locally

jcajka commented:
> @jlebon @r4f4 checking that out atm, locally

hm... it seems that the cosa's expected version of ignition is still skewed with the repos. @cgwalters @dustymabe can we get that fixed?

r4f4 commented:
https://github.com/jlebon/coreos-assembler/commit/0381a0a6a8592b57700db559771a91988a31fbc9 I guess.

sinnykumari commented:
> [jlebon@0381a0a](https://github.com/jlebon/coreos-assembler/commit/0381a0a6a8592b57700db559771a91988a31fbc9) I guess.

Yeah! This should be fixed soon, we have https://kojipkgs.fedoraproject.org/repos-dist/f29-coreos-continuous/latest/ repo available now which has latest ignition for all three arches.

jcajka commented:
@jlebon yup, fixed
@sinnykumari @r4f4 could you test it now? it should be working out of the box now.

jlebon commented:
> I think I'd prefer to just use the same approach for all architectures. This means we would default to creating a layered disk image and injecting the ignition config.

I'm not sure I agree with this. It's definitely nice to be consistent, but IMO this is outweighed by `cosa run` becoming much more intensive than it was. E.g. locally I went from 16s to the VM console to 34s with this patch. Might seem insignificant, but when iterating over e.g. Ignition configs, it adds up fast! :hourglass_flowing_sand: 

Will let others chime in, though I think I'd rather keep this conditional on architectures that don't support `qemu_fw_cfg` given that most developers do hack on x86_64. (And ideally, we'd have multi-arch CI to make sure we don't regress for the other arches).

jlebon commented:
Hmm, also note `cosa run` is used outside of the workdir contexts with `-d my-img.qcow2`, in which case, this now fails since it assumes `$PWD` is the workdir with `tmp/`.

This ties into Dusty's point above:

> It doesn't look like we are cleaning up the layered disk image after the VM is done booting. We should do that right?

An alternative approach here so that we keep the `exec qemu` pattern could be to open the layered disk, delete it, then pass the fd to qemu. E.g. something like:

```diff
diff --git a/src/cmd-run b/src/cmd-run
index 942470f..7403b99 100755
--- a/src/cmd-run
+++ b/src/cmd-run
@@ -91,7 +91,6 @@ fi

 LAYERED_DISK=tmp/layered-qemu.qcow2
 qemu-img create -b "$(realpath "${VM_DISK}")" -f qcow "$(realpath "${LAYERED_DISK}")"
-VM_DISK=${LAYERED_DISK}

 # Emulate the host CPU closely in both features and cores.
 set -- -machine accel=kvm -cpu host -smp "${VM_NCPUS}" "$@"
@@ -166,12 +165,17 @@ EOF
     fi
 fi

-coreos_gf_run_mount "${VM_DISK}"
+coreos_gf_run_mount "${LAYERED_DISK}"
 coreos_gf mkdir-p /boot/ignition
 coreos_gf upload ${IGNITION_CONFIG_FILE} /boot/ignition/config.ign
 #TODO coreos_gf_relabel /boot/ignition/config.ign
 coreos_gf_shutdown

+exec 3<> "${LAYERED_DISK}"
+exec 4< "${LAYERED_DISK}"
+rm -rf "${LAYERED_DISK}"
+VM_DISK=/dev/fdset/2,index=0,media=disk
+
 if [ -n "${SSH_PORT}" ]; then
    hostfwd=",hostfwd=tcp::${SSH_PORT}-:22"
 fi
@@ -193,4 +197,5 @@ exec ${QEMU_KVM} -name coreos -m "${VM_MEMORY}" -nographic \
               -netdev user,id=eth0,hostname=coreos"${hostfwd:-}" \
               -device virtio-net-pci,netdev=eth0 \
               -object rng-random,filename=/dev/urandom,id=rng0 -device virtio-rng-pci,rng=rng0 \
+              -add-fd fd=3,set=2 -add-fd fd=4,set=2 \
               "$@"
```

Then we can put the tmp image somewhere else, e.g. `/tmp/cosa-private` or something?

cgwalters commented:
> It doesn't look like we are cleaning up the layered disk image after the VM is done booting. We should do that right?

Being able to avoid exactly these types of resource leaks by `execve`ing qemu directly and passing it fds is an advantage we should indeed use here (and is a big class of problem introduced by using libvirt).

jcajka commented:
@dustymabe @cgwalters What is current stance on this PR, will it be merged if this change will be made across all arches?(taking in account @jlebon comment) I'm currently working on rewriting/forwardporting this taking in account the PR#507 and the feedback and it would really help me to avoid yet another rewrite/forwardport cycle.

bgilbert commented:
Adding another voice in favor of using libguestfs only on those architectures that need it.  Ultimately we should get qemu to support Ignition userdata on all arches; let's not embrace the libguestfs approach more broadly than needed for now.

dustymabe commented:
If the performance is drastically hurt by this (as suggested by @jlebon) then I understand keeping the approach separate for different architectures. As benjamin mentioned it would be nice to get a solution for other arches for ignition config injection.

dustymabe commented:
> @dustymabe @cgwalters What is current stance on this PR, will it be merged if this change will be made across all arches?

I think from the discussion above the desire is to keep it separate for now. Mind cleaning up this PR with that strategy in mind and we'll iterate on the review ?

jcajka commented:
@dustymabe Thanks for the clarification. I have been bit confused. I will enable the injection only on ppc64le and s390x. I will push new patch shortly, just finishing testing.

jcajka commented:
Sorry for delay. I have just pushed new iteration. Unfortunately I had some issues with getting HW for testing. It works for me on ppc64le. I would appreciate more testing on x86_64 and aarch64(not sure if fw_cfg is working there).

dustymabe commented:
> Sorry for delay. I have just pushed new iteration. 

Works for me!

> Unfortunately I had some issues with getting HW for testing. It works for me on ppc64le. I would appreciate more testing on x86_64 and aarch64(not sure if fw_cfg is working there).

worked for me on x86_64 - maybe we can get @sinnykumari to check aarch64

I'm +1 for merge.

cgwalters commented on src/cmd-run:
```diff
@@ -184,13 +172,38 @@ EOF
+    coreos_gf_run_mount "${VM_IMG}"
+    coreos_gf mkdir-p /boot/ignition
+    coreos_gf upload ${IGNITION_CONFIG_FILE} /boot/ignition/config.ign
+    #TODO coreos_gf_relabel /boot/ignition/config.ign
```
Interesting that we don't need to relabel.  I guess because we load this before SELinux policy is loaded?

cgwalters commented:
> I'm +1 for merge.

Simultaneously inserting my launch key too and turning...

dustymabe commented:
> > I'm +1 for merge.
> 
> Simultaneously inserting my launch key too and turning...

Yeah I was just saying when ready I've already given my go-ahead. 

> worked for me on x86_64 - maybe we can get @sinnykumari to check aarch64

^^ this was the part that possibly needed confirming before merge.

jlebon commented on src/cmd-run:
```diff
@@ -184,13 +172,38 @@ EOF
+else
+    echo "Re-using existing ${VM_PERSIST_IMG}"
+    VM_IMG=${VM_PERSIST_IMG}
+fi
```
Doesn't this new logic break `--persist-img` though? Since we're no longer doing `qemu-img create` on the first run anymore, right?

dustymabe replied:
yeah I guess so - TBH this logic kinda makes my head hurt. Would be nice to simplify it if possible or add comments where necessary.

jcajka replied:
Oh... so it is suposed to be managed by this code, I miss understood that part of code. Variable naming has been bit confusing. What do you want me to do about it? Another PR?

jcajka replied:
@dustymabe  reasons for this block is that gustfish doesn't like deleted files as backing storage. Other has been to rename the persistent image to some thing more descriptive, because the "persistent" one has been in most case just temporal and caused some confusion for me. It is bit unfortunate this got caught post review.

dustymabe replied:
@jcajka - yep this logic was needed. It just required me to read it a few times. Some comments would help I think.

sinnykumari commented:
Ran on aarch64, LGTM.
Note: Tested this with `setenforce 0` on aarch64 host because of "error while loading shared libraries: libtinfo.so.6: cannot change memory protections" while doing `podman run/build` similar to issue described [here](https://github.com/containers/libpod/issues/1866#issuecomment-441697053)
